R/convert.R

```diff
@@ -20,20 +35,26 @@
 #' foo1 <- new_class("foo1", properties = list(x = class_integer))
 #' foo2 <- new_class("foo2", foo1, properties = list(y = class_double))
 #'
-#' method(convert, list(foo1, class_integer)) <- function(from, to) from@x
-#' method(convert, list(foo2, class_double)) <- function(from, to) from@y
+#' # S7 provides a default implementation for coercing an object to one of
+#' # its parent classes:
+#' convert(foo2(x = 1L, y = 2), to = foo1)
 #'
-#' convert(foo1(x = 1L), to = class_integer)
-#' try(convert(foo1(x = 1L), to = class_double))
+#' # For all other cases, you'll need to provide your own.
+#' try(convert(foo1(x = 1L), to = class_integer))
 #'
-#' convert(foo2(x = 1L, y = 2), to = class_integer)
-#' convert(foo2(x = 1L, y = 2), to = class_double)
-#' convert(foo2(x = 1L, y = 2), to = foo1)
+#' method(convert, list(foo1, class_integer)) <- function(from, to) {
+#'   from@x
+#' }
+#' convert(foo1(x = 1L), to = class_integer)
 #'
-#' # If we define a convert method for integer + foo1:
-#' method(convert, list(class_integer, foo1)) <- function(from, to) foo1(x = from)
+#' # Note that conversion does not respect inheritance so if we define a
+#' # convert method for foo1 to integer:
+#' method(convert, list(class_integer, foo1)) <- function(from, to) {
+#'   foo1(x = from)
+#' }
 #' convert(1L, to = foo1)
-#' # Converting too foo2 still errors
+#'
+#' # Converting too foo2 will still error
```

Comment:
This example actually seems somewhat obvious to me, because we don't have `y`, which foo2 needs, so of course it fails!

In my head a more interesting example could be to have

```
bar <- new_class("bar", properties = list(x = class_integer, y = class_double))
```

If you define a `convert()` method for `bar -> foo2`, do you also get `bar -> foo1` "for free"? This would be theoretically possible because `foo1` only needs `x`, and you could get there by "automatically" chaining explicit convert methods like `bar -> foo2 -> foo1`, but I don't think we do this (nor should we)

Reply:
I think this is the most common failure more because most subclasses will add additional properties, so this is the common failure mode. Maybe I can just explain that.